src/iosrv/obj.c

```diff
@@ -38,24 +36,139 @@
 #include <daos_task.h>
 
 #include <daos_srv/daos_server.h>
+#include "srv_internal.h"
+
+/*
+ * The main reason we want to separate this DSC progress ULT with the hardware
+ * progress ULT (where the network and NVMe are polled) is that the DSC task
+ * completion callback and task run (usually on retry) calls client APIs which
+ * may yield on grabbing some global locks.
```

Comment:
I ever considered to remove this progress ULT (original dss_tse_progress_ult), seems it should work by calling  tse_sched_progress() in the main service progress loop in dss_srv_handler().

I did not quite understand above comment about "calls client APIs which may yield on grabbing some global locks", can you give an example? and is there any difference with just calling tse_sched_progress() in the main service progress loop in dss_srv_handler()?

Reply:
Look into the dsc_obj_retry_cb() (which will be called directly in the progress ULT) and the task function (DAOS client APIs, usually called on retry in this progress ULT), you'll see sometimes we need to to grab some global locks (like pool map lock), that'll make the progress ULT yield and stop hardware polling unintentionally. That'll slow down hardware polling and it's deadlock prone. (imaging what if the lock holder waiting on a hardware event, like RPC or NVMe IO completion)

In a word, I think the main progress ULT (dss_srv_handler()) should never yield and waiting on something (lock, or some other events).

I actually think that instead of relying on 'tse' (the client asynchronous engine which looks not quite scalable), we'd create ULT for these kind of DSC calls, and rely on scalable Argobots to schedule them, but that requires lots of changes to client API to get rid of the dependency on 'tse', so I don't plan to do that in this patch.

Reply:
The reason we use separate progress is because this is only needed during rebuild.  If you want to explore this for your own case,  probably ds_obj_xxx() should start a new ULT for remote fetch/iteration, instead of putting itself to the scheduler.

"that'll make the progress ULT yield and stop hardware polling unintentionally. "

Not sure I follow here, even if these global locks is not in the crt(nvme) progress ULT, if it is not yield during the lock (i.e. not using ABT lock), it will block crt(nvme) progress anyway. 

So probably using client stack on server, which does not use ABT lock, may not be good here.  So we probably should put this remote fetch/iteration on a different xstream. 

Btw: could you please explain why the faulty detection needs this? does it need task/scheduler?

Reply:
> Not sure I follow here, even if these global locks is not in the crt(nvme) progress ULT, if it is not yield during the lock (i.e. not using ABT lock), it will block crt(nvme) progress anyway.

I think you are right, if it's a global pthread lock (not ABT lock) on client stack, the whole xstream could be blocked. That's a problem of calling client API on server side, any idea to solve this? (Re-implement client API for server looks not quite practical)

Reply:
> Btw: could you please explain why the faulty detection needs this? does it need task/scheduler?
Faulty detection needs to call pool target exclude asynchronously on server side.